drivers/platform/mbed/mbed_uart.cpp

```diff
@@ -0,0 +1,492 @@
+		return -EINVAL;
+
+	if (!((struct mbed_uart_desc *)(desc->extra))->uart_port)
+		return -EINVAL;
+
+	if (((struct mbed_uart_desc *)desc->extra)->virtual_com_enable) {
+		usb_cdc_dev = (platform_usbcdc *)((struct mbed_uart_desc *)(
+				desc->extra))->uart_port;
+
+		while (bytes_number) {
+			while (!usb_cdc_dev->data_transmited()) {
+				/* Wait until old data is transmitted */
+			}
+
+			/* Make sure packet size is less than max CDC packet size during data transmit */
+			d_sz = (bytes_number > (USB_CDC_MAX_PACKET_SIZE - 1)) ?
+			       (USB_CDC_MAX_PACKET_SIZE - 1) :
+			       bytes_number;
+
+			/* Change terminal connection status manually */
+			usb_cdc_dev->change_terminal_connection(true);
+
+			usb_cdc_dev->send_nb((uint8_t *)&data[indx], d_sz, &d_sz);
+
+			bytes_number -= d_sz;
+			indx += d_sz;
+		}
+
+		return tx_bytes;
+	} else {
+		uart = (BufferedSerial *)(((struct mbed_uart_desc *)(
+						   desc->extra))->uart_port);
+		return uart->write(data, bytes_number);
+	}
+}
+
+/**
+ * @brief Submit reading buffer to the UART driver.
+ * Buffer is used until bytes_number bytes are read.
+ * @param desc[in]- Descriptor of the UART device
+ * @param data[out] - Buffer where data will be read
+ * @param bytes_number[in] - Number of bytes to be read.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ * @note Currently implemented only for UART and not for USBCDC (VCOM)
+ */
+int32_t uart_read_nonblocking(struct uart_desc *desc,
+			      uint8_t *data,
+			      uint32_t bytes_number)
+{
+	mbed::BufferedSerial *uart;	// pointer to BufferedSerial/UART instance
+	uint32_t i;
+
+	if (!desc || !desc->extra || !data)
+		return -EINVAL;
+
+	if (!((struct mbed_uart_desc *)(desc->extra))->uart_port)
+		return -EINVAL;
+
+	uart = (BufferedSerial *)(((struct mbed_uart_desc *)(
+					   desc->extra))->uart_port);
+
+	for (i = 0; i < bytes_number; i++) {
+		if (uart->readable() > 0)
+			uart->read(&data[i], 1);
+	}
+
+	return bytes_number;
+}
+
+/**
+ * @brief Submit writting buffer to the UART driver.
+ * Data from the buffer is sent over the UART, the function returns immediatly.
+ * @param desc[in] - Descriptor of the UART device
+ * @param data[in,out] - Buffer where data will be written
+ * @param bytes_number[in] - Number of bytes to be written.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ * @note Currently implemented only for UART and not for USBCDC (VCOM)
+ */
+int32_t uart_write_nonblocking(struct uart_desc *desc,
+			       const uint8_t *data,
+			       uint32_t bytes_number)
+{
+	mbed::BufferedSerial *uart;		// pointer to BufferedSerial/UART instance
+	uint32_t i;
+
+	if (!desc || !desc->extra || !data)
+		return - EINVAL;
+
+	if (!((struct mbed_uart_desc *)(desc->extra))->uart_port)
+		return -EINVAL;
+
+	uart = (BufferedSerial *)(((struct mbed_uart_desc *)(
+					   desc->extra))->uart_port);
+
+	for (i = 0; i < bytes_number; i++) {
+		if (uart ->writable() > 0)
```

Comment:
Same as the read: we should query once at the beginning if it's writable, and, if yes, `uart->write(data, bytes_number);`. There may be some checks here that we can actually transfer that many.